File: declarative-android-hn/app/src/main/java/com/prof18/declarativeandroidhn/ui/MainViewModel.kt
package com.prof18.declarativeandroidhn.ui

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.prof18.declarativeandroidhn.data.model.AppState
import com.prof18.declarativeandroidhn.data.model.NewsState
import com.prof18.declarativeandroidhn.data.newsList
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class MainViewModel : ViewModel() {

    private val _appState = MutableLiveData<AppState>(AppState(NewsState.Loading))
    val appState: LiveData<AppState>
        get() = _appState

    init {
        generateError()
    }

    fun loadData() {
        viewModelScope.launch {
            _appState.value = AppState(NewsState.Loading)
            delay(2000)
            _appState.value = AppState(NewsState.Success(newsList))
        }
    }

    private fun generateError() {
        viewModelScope.launch {
            delay(2000)
            _appState.value =
                AppState(NewsState.Error("This is a generated error only to try an error state"))
        }
    }
}

File: declarative-ios-hn/Declarative iOs HN/ui/MainViewModel.swift
//
//  MainViewModel.swift
//  Declarative iOs HN
//
//  Created by Marco Gomiero on 14/02/2020.
//  Copyright © 2020 Marco Gomiero. All rights reserved.
//

import Foundation
import SwiftUI
import Combine

class MainViewModel: ObservableObject {
    
    @Published private(set) var appState: AppState = AppState()
    
    func loadData() {
        self.appState = AppState(newsState: Loading())
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
            self.appState = AppState(newsState: Success(news: newsList))
        }
    }
    
    func generateError() {
        self.appState = AppState(newsState: Loading())
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
            self.appState = AppState(newsState: Error(reason: "This is a generated error only to try an error state"))
        }
    }
}
